pkg/kclient/generator/generators.go

```diff
@@ -0,0 +1,347 @@
+		}
+	}
+
+	return container
+}
+
+// GetContainers iterates through the components in the devfile and returns a slice of the corresponding containers
+func GetContainers(devfileObj devfileParser.DevfileObj) ([]corev1.Container, error) {
+	var containers []corev1.Container
+	for _, comp := range GetDevfileContainerComponents(devfileObj.Data) {
+		envVars := convertEnvs(comp.Container.Env)
+		resourceReqs := getResourceReqs(comp)
+		ports := convertPorts(comp.Container.Endpoints)
+		containerParams := ContainerParams{
+			Name:         comp.Name,
+			Image:        comp.Container.Image,
+			IsPrivileged: false,
+			Command:      comp.Container.Command,
+			Args:         comp.Container.Args,
+			EnvVars:      envVars,
+			ResourceReqs: resourceReqs,
+			Ports:        ports,
+		}
+		container := GenerateContainer(containerParams)
+
+		// If `mountSources: true` was set, add an empty dir volume to the container to sync the source to
+		// Sync to `Container.SourceMapping` and/or devfile projects if set
+		if comp.Container.MountSources {
+			syncRootFolder := addSyncRootFolder(container, comp.Container.SourceMapping)
+
+			err := addSyncFolder(container, syncRootFolder, devfileObj.Data.GetProjects())
+			if err != nil {
+				return nil, err
+			}
+		}
+		containers = append(containers, *container)
+	}
+	return containers, nil
+}
+
+// PodTemplateSpecParams is a struct that contains the required data to create a pod template spec object
+type PodTemplateSpecParams struct {
+	ObjectMeta     metav1.ObjectMeta
+	InitContainers []corev1.Container
+	Containers     []corev1.Container
+	Volumes        []corev1.Volume
+}
+
+// GeneratePodTemplateSpec creates a pod template spec that can be used to create a deployment spec
+func GeneratePodTemplateSpec(podTemplateSpecParams PodTemplateSpecParams) *corev1.PodTemplateSpec {
+	podTemplateSpec := &corev1.PodTemplateSpec{
+		ObjectMeta: podTemplateSpecParams.ObjectMeta,
+		Spec: corev1.PodSpec{
+			InitContainers: podTemplateSpecParams.InitContainers,
+			Containers:     podTemplateSpecParams.Containers,
+			Volumes:        podTemplateSpecParams.Volumes,
+		},
+	}
+
+	return podTemplateSpec
+}
+
+// DeploymentSpecParams is a struct that contains the required data to create a deployment spec object
+type DeploymentSpecParams struct {
+	PodTemplateSpec   corev1.PodTemplateSpec
+	PodSelectorLabels map[string]string
+}
+
+// GenerateDeploymentSpec creates a deployment spec
+func GenerateDeploymentSpec(deployParams DeploymentSpecParams) *appsv1.DeploymentSpec {
+	deploymentSpec := &appsv1.DeploymentSpec{
+		Strategy: appsv1.DeploymentStrategy{
+			Type: appsv1.RecreateDeploymentStrategyType,
+		},
+		Selector: &metav1.LabelSelector{
+			MatchLabels: deployParams.PodSelectorLabels,
+		},
+		Template: deployParams.PodTemplateSpec,
+	}
+
+	return deploymentSpec
+}
+
+// GeneratePVCSpec creates a pvc spec
+func GeneratePVCSpec(quantity resource.Quantity) *corev1.PersistentVolumeClaimSpec {
+
```

Comment:
devfile should be input here and generate pvc specs according to volume component

Reply:
I left this untouched due to the GetVolume refactoring I was talking about. CreatePVC just takes `corev1.PersistentVolumeClaimSpec` spec to create the pvc. The pvc name is tied with other logic and would need a separate PR to refactor it. I think thats going to be a bigger change.